proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,291 @@
+distribution block during the reward distribution phase, i.e.
+`address = hash("SysvarRewardReserve11111111111111111111111111111", block_height)`.
+
+
+```
+crate::declare_sysvar_id!("SysvarRewardReserve11111111111111111111111111111", EpochRewardReserve);
+
+struct EpochRewardReserve {
+   reserve_hash: Hash,
+}
+```
+
+The account balance of `RewardsReserve` is equal to the total amount (in
+lamports) of all rewards to be distributed in a block. The `reserve_hash` is
+computed from the rewards in the block as follows.
+
+```
+reserve_hash = hash([(stake_account_key, reward_amount)])
+```
+
+At the end of reward computation phase, M `EpochRewardsReserve` sysvar
+accounts are updated, where `M` corresponds to how many blocks will be in
+the reward distribution phase, which is described in the next section.
+
+
+### Reward Distribution
+
+Reward distribution phase happens right after reward computation phase and
+lasts for `M` blocks. Each of the `M` blocks is responsible for distributing
+the reward from one partition of the rewards stored in the `EpochRewardReserve`
+account, which address
+`hash("SysvarRewardReserve11111111111111111111111111111", block_height)`.
+
+During the reward distribution phase, for the block at `block_height`, the
+rewards will be distributed from the `EpochRewardReserve` sysvar account with
+the address of "hash(`SysvarRewardReserve11111111111111111111111111111111`,
+block_height). As the stake_accounts receive their rewards from the reserve
+account, the balance of the EpochRewardReserve will decrease. Once all rewards
+have been distributed, the balance of the EpochRewardReserve account should be
+reduced to 0.
+
+Before each reward distribution, the `EpochRewardReserve` account's balance
+is checked to make sure there is enough balance to distribute the reward. After a
+reward is distributed, the balance in `EpochRewardReserve` account is decreased
+by the amount of the reward that was distributed. The hash of reward_account
+pubkey and reward amount is accumulated when rewards are distributed. After all
+the rewards are distributed from the reserve account for the block, the hash
+computed from all the rewards distribution is verified against the hash stored
+in the `EpochRewardReserve` account to make sure they match.
+
+During the reward distribution phase, there is a "root accumulator hash" that
+starts empty and is gradually filled by adding each hash of a
+`EpochRewardReserve`. The distribution process continues until the "root
+accumulator hash" matches the "root hash" stored in `EpochRewardHistory`. At this
+point, the reward distribution for the current epoch is complete and we can
+safely assume that there will be no further distributions for the remainder of
+the epoch.
+
+### Restrict Stake Account Access
+
+To avoid the complexity and potential rewards in unexpected stake accounts, all
+write access to stake accounts are restricted. That means any update to stake
```

Comment:
I don't think we need to restrict all write access to stake account actually. For example credit of lamports to a stake account would not be a problem.

Instead we had discussed was disabling the stake program entirely from the start of an epoch through the completion of rewards distribution, where any attempt to invoke the stake program results in a new unique error code like "StakeProgramUnavailable".